### PR TITLE
[JENKINS-65813] Add @Symbol to support pipeline

### PR DESCRIPTION
**Pull request by mamh2021**
Signed-off-by: bright.ma RMSH06 <bright.ma@blackshark.com>

### CODE REVIEW

**Review by kinow — COMMENTED** [review]
Hi @mageSFC2020 thanks for the PR. I just need to test and confirm it works with no issues, but looks good from a first quick review. I am not sure what names/symbols would be best to describe each parameter, but we can discuss that later :+1: 

Ref: https://www.jenkins.io/doc/developer/plugin-development/pipeline-integration/#defining-symbols

File: src/main/java/org/biouno/unochoice/DynamicReferenceParameter.java [hunk 1]
```diff
@@ -131,6 +132,7 @@ public String getChoicesAsStringForUI() {
     // --- descriptor
 
     @Extension
+    @Symbol({"referenceChoice"})
```
Comment (kinow) [inline src/main/java/org/biouno/unochoice/DynamicReferenceParameter.java:135]: @imoutsatsos what do you think of these three aliases for pipeline users?

- `activeChoice` -> Active Choices Parameter
- `referenceChoice` -> Active Choices Reactive Reference Parameter
- `reactiveChoice` -> Active Choices Reactive Parameter

I think if we were choosing parameter names today, knowing about this pipelines aliases thing, probably we would give the parameters other names to make it easier to pick up aliases :-) but hindsight is always 20/20 right.

**Review by kinow — COMMENTED** [review]
Pull request changes look really good. I'll create a JIRA and update the PR to track it. I've read the https://www.jenkins.io/doc/developer/plugin-development/pipeline-integration page, and other than what's here, I think the only suggestion I'd have is to add an explicit dependency to `org.jenkins-ci.plugins:structs` since the documentation states

> To make for a more attractive and mnemonic usage style, you can depend on org.jenkins-ci.plugins:structs and add a @Symbol to your Descriptor, uniquely identifying it among extensions of its kind (in this example, SimpleBuildSteps):

I'm looking at other issues from users who are using the plug-in with Pipelines (I don't use them). Once I've gotten a better grasp of how these Symbols will work, have tested them, and @imoutsatsos has had time to comment on my question for him, this should be ready to be merged.

I'll check whether it's possible to write a small junit test for it too, but not a blocker for the PR IMO.

Thanks @mamh2021 !

**Comment by kinow** [comment]
Placeholder ticket: https://issues.jenkins.io/browse/JENKINS-65813

**Comment by mawinter69** [comment]
Any idea when you will merge this?
It is really a pain to configure active choice parameters in a Jenkinsfile when you can't use the declarative directive generator.
With this change you can use it and it will generate you this:

```
parameters {
  activeChoice choiceType: 'PT_RADIO', description: 'test', filterLength: 1, filterable: true, name: 'test', randomName: 'choice-parameter-1322655067846700', script: groovyScript(fallbackScript: [classpath: [], oldScript: '', sandbox: false, script: 'return ["Error"]'], script: [classpath: [], oldScript: '', sandbox: false, script: 'return ["abc", "def", "ghi"]'])
}

```

**Comment by kinow** [comment]
>Any idea when you will merge this?

Hi @mawinter69 ,

I don't have a roadmap as I am not currently using the plug-in. I'm co-maintaining with @imoutsatsos and working on the features I can when I have spare time to volunteer.

I had a look at the comments, and looks like it simply lost momentum. I think I had had the docs about the `@Symbol` support in Jenkins, and only had some questions about the names used, and also needed to do some quick testing before reviewing/merging.

Maybe if someone else takes over and answer the comments, and then confirms this PR works, that would speed up the release of this feature.

-Bruno

**Comment by mawinter69** [comment]
> Pull request changes look really good. I'll create a JIRA and update the PR to track it. I've read the https://www.jenkins.io/doc/developer/plugin-development/pipeline-integration page, and other than what's here, I think the only suggestion I'd have is to add an explicit dependency to `org.jenkins-ci.plugins:structs` since the documentation states
> 
> > To make for a more attractive and mnemonic usage style, you can depend on org.jenkins-ci.plugins:structs and add a @symbol to your Descriptor, uniquely identifying it among extensions of its kind (in this example, SimpleBuildSteps):
> 

The @Symbol  annotation is not declared in the structs plugin. You need the structs plugin only when you need to lookup things that are annotated with symbols which is not the case here.

**Comment by imoutsatsos** [comment]
@mawinter69  , @kinow Thanks for the PR and review! I don't use Jenkins pipeline files for my projects and so I can't comment beyond the proposed changes for the names while using the directive generator. Nonetheless I fully support the evolution of the Active Choices plugin to support pipeline files as this is what most DevOp Jenkins users have migrated to. 
The following suggestions seem appropriate:

`activeChoice` -> Active Choices Parameter
`activeChoiceHtml` -> Active Choices Reactive Reference Parameter
`reactiveChoice` -> Active Choices Reactive Parameter

I prefer the `activeChoiceHtml` as it clarifies (but still maintains) the original name. I suggest not using reactive, as this parameter can be Groovy generated static HTML, without dependency on other parameters.

**Comment by mawinter69** [comment]
> 
> `activeChoice` -> Active Choices Parameter 
> `activeChoiceHtml` -> Active Choices Reactive Reference Parameter 
> `reactiveChoice` -> Active Choices Reactive Parameter
> 

I think these names are good and describe well what they do. If you can do a change with these names that would be really great as it eases use of the plugin in pipelines a lot.

**Comment by kinow** [comment]
> > `activeChoice` -> Active Choices Parameter
> > `activeChoiceHtml` -> Active Choices Reactive Reference Parameter
> > `reactiveChoice` -> Active Choices Reactive Parameter
> 
> I think these names are good and describe well what they do. If you can do a change with these names that would be really great as it eases use of the plugin in pipelines a lot.

I tried to rebase the branch of OP, but when opening the PR the option to allow contributors to modify the branch was not checked.

```bash
$ git push -f mamh-java pr20210225
ERROR: Permission to mamh-java/active-choices-plugin.git denied to kinow.
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.
```

So the only option is to start a new branch & pull request. I'm planning to cut a release with all the changes we have now on `master` this week, @mawinter69 

If we create a new PR and you or somebody else is able to check out the branch locally, build and test it, then we can include this PR in this release. Otherwise it will be for the next release some other day.

-Bruno

**Comment by mawinter69** [comment]
I can check the PR

**Comment by kinow** [comment]
Moved to #80 so that the branch can be updated.

**Pull request closed without merging**